pkg/bb/bbmain/cmd/main.go

```diff
@@ -27,7 +28,93 @@ func main() {
+// #!/bin/rc
+// bb ls
+// This leaves a lot to be desired: it puts the execution of a shell in front
+// of each u-root command, and it requires the existence of that shell on the
+// system.
+//
+// The goal is that a single u-root file lead to running the u-root busybox
+// with no intermediate programs running.
+//
+// It is worth taking a look at what a symlink is, how it works in operation,
+// and how we might achieve the same goal some other way.
+//
+// A symlink is plain file, containing 0 or more bytes of text (or utf-8, depending)
+// with an attribute that causes the kernel to give it special treatment.
+// It is not available on all file systems.
+//
+// [Note: they were invented in 1965 for Multics].
+// The symlink is itself still controversial, though widely used.
+//
+// Consider the process of traversing a symlink: it involves the equivalent
+// of stat, open, read, evaluate contents, use that as a file name, repeat as needed.
+//
+// It is possible to get that same effect, with the same overheads, by using #!
+// files but specifying bb as the interpreter.
+//
+// ls would then be:
+// #!/bin/bb ls
+//
+// Note that the absolute path is required, else Linux will throw an error as bb
+// is not in the list of allowed interpreters.
+// The /bin/bb path is not an issue on Plan 9, since users construct their name space
+// on startup and binding /bbin into /bin is no problem.
+//
+// In this case the kernel will stat, open, and read the file, find the executable name,
+// and start it. This approach has as low overhead as the symlink approach.
+//
+// One problem remains: Unix and Plan 9 evaluate arguments in a #! file differently,
+// and, further, invoke the argument in a different way.
+// Given the file shown above, bb on Linux gets the arguments:
+// [ls ls /tmp/ls]
+// With the same file, bb on linux gets this:
+// [/bbin/bb ls /tmp/ls]
+// But wait! There's more!
+// On Plan 9, the arguments following the interpreter are tokenized (split on space)
+// and on Linux, they are not.
+//
+// This leads to a few conclusions:
+// - We can get around lack of symlinks by using #! (sh-bang) files with an absolute path to
+//   bb as the interpreter, e.g. #!/abs/path/to/bb argument.
+//   This achieves the "exec once" goal.
+// - We can specify which u-root tool to use via arguments to bb in the #! file.
+// - The argument to the interpreter (/bbin/bb) should be one token (e.g. ls) because of different
+//   behavior in different systems (some tokenize, some do not).
+// - Because of the differences in how arguments are presented to #! on different kernels,
+//   there should be a reasonably unique marker so that bb can have confidence that
+//   it is running as an interpreter.
+//
+// The conclusions lead to the following design:
+// #! files for bb specify their argument with #!. E.g., the file for ls looks like this:
+// #!/bbin/bb #!ls
+// On Linux, the args to bb then look like:
+// [/tmp/u/bbin/bb #!ls /tmp/ls ...]
```

Comment:
where did the /tmp/ls come from?

#!/bbin/bb ls turns into [/bbin/bb ls], no? Where does the third arg come from?

Reply:
I am not sure why it does this, but that's what it does ...I just had bb dump its args.

But I need to correct that text, while I was testing this, I was using /tmp/u/bbin/bb

Reply:
ah right fhs explained it:

i think it's correct. The command after #! is assumed to be an interpreter which takes the script as the first argument. So, for rc scripts, it runs /bin/rc /path/to/script arg1 arg2 ... . There is no other way to pass in the arguments -- you can't do /bin/rc arg1 arg2 < /path/to/script

so the kernel passes in the argument as it was made to the interpreter (ls) but also adds in the full path that got it there (/tmp/ls) -- basically you get  3 pieces of info BEFORE the args to the command: interpreter, arg in the #! file, and full path of the file that is the #! file.

Reply:
and note, plan 9 arguably gives you less: 
[basename arg fullpath]
i.e.
[ls ls /tmp/ls]